core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -240,186 +202,60 @@ protected IndexShard getShard(ShardId shardId) {
+    class ShardsIndicesStatusChecker implements Runnable {
 
-            if (changes.isEmpty() == false) {
-                // Something changed: recompute indexing buffers:
-                calcAndSetShardBuffers("[" + changes + "]");
-            }
-        }
+        @Override
+        public synchronized void run() {
 
-        /**
-         * goes through all existing shards and check whether there are changes in their active status
-         */
-        private void updateShardStatuses(EnumSet<ShardStatusChangeType> changes) {
+            // Fast check to sum up how much heap all shards' indexing buffers are using now:
+            long totalBytesUsed = 0;
             for (ShardId shardId : availableShards()) {
 
-                // Is the shard active now?
-                Boolean isActive = getShardActive(shardId);
+                // Give shard a chance to transition to inactive so sync'd flush can happen:
+                checkIdle(shardId, inactiveTime.nanos());
 
-                if (isActive == null) {
-                    // shard was closed..
-                    continue;
-                }
+                totalBytesUsed += getIndexBufferRAMBytesUsed(shardId);
+            }
 
-                // Was the shard active last time we checked?
-                Boolean wasActive = shardWasActive.get(shardId);
-
-                if (wasActive == null) {
-                    // First time we are seeing this shard
-                    shardWasActive.put(shardId, isActive);
-                    changes.add(ShardStatusChangeType.ADDED);
-                } else if (isActive) {
-                    // Shard is active now
-                    if (wasActive == false) {
-                        // Shard became active itself, since we last checked (due to new indexing op arriving)
-                        changes.add(ShardStatusChangeType.BECAME_ACTIVE);
-                        logger.debug("marking shard {} as active indexing wise", shardId);
-                        shardWasActive.put(shardId, true);
-                    } else if (checkIdle(shardId, inactiveTime.nanos()) == Boolean.TRUE) {
-                        // Make shard inactive now
-                        changes.add(ShardStatusChangeType.BECAME_INACTIVE);
-                        logger.debug("marking shard {} as inactive (inactive_time[{}]) indexing wise",
-                                     shardId,
-                                     inactiveTime);
-                        shardWasActive.put(shardId, false);
+            if (totalBytesUsed > indexingBuffer.bytes()) {
+                // OK we are using too much; make a queue and ask largest shard(s) to refresh:
+                logger.debug("now refreshing some shards: total indexing bytes used [{}] vs index_buffer_size [{}]", new ByteSizeValue(totalBytesUsed), indexingBuffer);
+                PriorityQueue<ShardAndBytesUsed> queue = new PriorityQueue<>();
+                for (ShardId shardId : availableShards()) {
```

Comment:
is it possible that if we can't refresh quick enough that we schedule a large amount of refreshes even if one is still _pending_? if the node is busy it can pile up quite a bit no? I think we should have at most one refresh in flight per index shard otherwise we might refresh the same shard over and over again? 

When we worked on DWPT you pushed hard on having some stall control that blocks indexing threads if we can't flush fast enough, do we need something like this here as well? over paranoid? ;)